server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -279,49 +255,28 @@ public ShardRouting[] started() {
      * @return a list of shards
      */
     public List<ShardRouting> shardsWithState(String index, ShardRoutingState... states) {
-        List<ShardRouting> shards = new ArrayList<>();
-
-        if (states.length == 1) {
-            if (states[0] == ShardRoutingState.INITIALIZING) {
-                for (ShardRouting shardEntry : initializingShards) {
-                    if (shardEntry.getIndexName().equals(index) == false) {
-                        continue;
-                    }
-                    shards.add(shardEntry);
-                }
-                return shards;
-            } else if (states[0] == ShardRoutingState.RELOCATING) {
-                for (ShardRouting shardEntry : relocatingShards) {
-                    if (shardEntry.getIndexName().equals(index) == false) {
-                        continue;
-                    }
-                    shards.add(shardEntry);
-                }
-                return shards;
-            } else if (states[0] == ShardRoutingState.STARTED) {
-                for (ShardRouting shardEntry : startedShards) {
-                    if (shardEntry.getIndexName().equals(index) == false) {
-                        continue;
-                    }
-                    shards.add(shardEntry);
-                }
-                return shards;
-            }
-        }
+        return Stream.of(states).flatMap(state -> shardsWithState(index, state).stream()).collect(toCollection(ArrayList::new));
+    }
 
-        for (ShardRouting shardEntry : this) {
-            if (shardEntry.getIndexName().equals(index) == false) {
-                continue;
-            }
-            for (ShardRoutingState state : states) {
-                if (shardEntry.state() == state) {
-                    shards.add(shardEntry);
-                }
+    public List<ShardRouting> shardsWithState(String index, ShardRoutingState state) {
+        var shards = new ArrayList<ShardRouting>();
+        for (ShardRouting shardEntry : getShardsWithState(state)) {
+            if (shardEntry.getIndexName().equals(index)) {
+                shards.add(shardEntry);
             }
         }
         return shards;
     }
 
+    private LinkedHashSet<ShardRouting> getShardsWithState(ShardRoutingState state) {
```

Comment:
IMO it's confusing to have `getShardsWithState` and `shardsWithState` which do quite different things (in terms of copying the internal value or not). How about naming this something like `internalGetShardsWithState`?